Fix instability in dragging and text selection while touchpad button is pressed

lvs1974 opened the pull request:
Fix instability in dragging and text selection while touchpad button is pressed and finger is re-touching for moving (dragging can be spontaneously cancelled, selection can start from another place).
To be more precise, I put links to video files:
[strange behaviour](https://drive.google.com/file/d/1w-rFoDpL_fDOUShz3WM4HIvjJc1ug-VM/view?usp=sharing)
[normal behaviour](https://drive.google.com/file/d/1XO_cVNQNAABF6J3735QoH6UIYuPBLj0S/view?usp=sharing)

vit9696 commented on VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp:
```diff
@@ -35,7 +35,7 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
         return;
     
-    IOLog("VoodooInput got input, count: %d\n", multitouch_event.contact_count);
+    //IOLog("VoodooInput got input, count: %d\n", multitouch_event.contact_count);
```
This is now removed.

vit9696 commented on VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp:
```diff
@@ -52,7 +52,7 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
     
     // physical button
-    input_report.Button = transducer->isPhysicalButtonDown;
+    UInt8 Button = transducer->isPhysicalButtonDown;
```
Probably best to make it lowercase.

vit9696 commented:
Sounds ok to me, but I am not really familiar with the code to be able to perform a decent review.

CC @kprinssu, @usr-sse2, @Goshin

kprinssu commented on VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp:
```diff
@@ -52,7 +50,7 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
     
     // physical button
-    input_report.Button = transducer->isPhysicalButtonDown;
+    UInt8 button = transducer->isPhysicalButtonDown;
```
This can be a simple `bool` rather than a `UInt8`.

kprinssu commented on VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp:
```diff
@@ -91,14 +89,28 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
         }
+    }
+    
+    for (int i = 0; i < multitouch_event.contact_count + 1; i++) {
```
Why the second inner loop?

lvs1974 replied:
@kprinssu: in the first loop we calculate input_active through all fingers.
In the second loop we use input_active disregarding for which finger input_active was detected.

usr-sse2 commented:
What gesture are you fixing? Is it dragging with physical button pressed? If so, then which button does your touchpad have – separate from the touch surface or the touch surface itself is clickable (Lenovo seems to call such touchpads ClickPads)?

lvs1974 commented:
> What gesture are you fixing? Is it dragging with physical button pressed? If so, then which button does your touchpad have – separate from the touch surface or the touch surface itself is clickable (Lenovo seems to call such touchpads ClickPads)?

My touch surface is clickable.
I am fixing dragging with one finger stay pressed on surface (in click state) and another finger is dragging something.
Text selection - very similar.

ben9923 commented:
The log fix you have there is still causing a conflict - you have an empty line while master has indentation. You might want to squash (probably fixups) & rebase.

Goshin commented:
@lvs1974 Hi, I cannot replicate the bug with VoodooI2C+VoodooInput. Would you mind elaborating on the cause of this bug you think?

The fix seems to apply the pressure value only to the first finger in the report, which makes sense to me, but I don't understand how it is related to the `input_active`.

There is another chance that the bug is due to the mess of the finger id. Are you using the VoodooPS2 or VoodooI2C? When it doesn't actually track the finger, the finger id of the first finger might change every time the second finger is taking off or retouching, which acts just like you release and repress the first finger at that moment. It is pretty consistent with the video as it only happens occasionally when the second finger takes off.

lvs1974 commented:
@Goshin, I don't think it can be reproduced on any hardware.
My touchpad has only I2C interface, so not PS2 kexts (except keyboard).
I added log statement in the beginning of the method:
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1540 (time - just a counter).
And I added log statements into the first loop, and below you can see, that in the beginning "Input active" is detected for both transducers 0 and 1, and button is clicked only for 0.
But it has changed later after a few movements: "Input active" is not detected for transducer 0, only for transducer 1, but clicked button is still detected for transducer 0.
I am not expert in touchpad input & hardware, I just added log statements, observed,  and this gave me a clue how I can fix it.

`VoodooInputSimulatorDevice::constructReportGated: enter, time = 1531
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1532
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1533
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1534
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1535
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1536
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1537
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1538
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1539
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1540
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1541
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1542
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1543
VoodooInputSimulatorDevice Input active is detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1544
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1545
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1546
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1547
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1548
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1549
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1550
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1551
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1552
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1553
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1554
VoodooInputSimulatorDevice Input active is not detected at 0, button = 0
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1555
VoodooInputSimulatorDevice Input active is not detected at 0, button = 0
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1556
VoodooInputSimulatorDevice Input active is not detected at 0, button = 0
VoodooInputSimulatorDevice Input active is detected at 1, button = 0`

lvs1974 commented:
@Goshin, below there is a log excerpt with finger_id (just in case):
`VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28949
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28950
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28951
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28952
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28953
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28954
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28955
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28956
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28957
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28958
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28959
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28960
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28961
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28962
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28963
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28964
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28965
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28966
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28967
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28968
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28969
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28970
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28971
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28972
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28973
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28974
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28975
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28976
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28977
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28978
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28979
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28980
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28981
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28982
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28983
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28984
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28985
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28986
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28987
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28988
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28989
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28990
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28991
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28992
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28993
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28994
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28995
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28996
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 28997
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 28998
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 28999
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 29000
VoodooInputSimulatorDevice::constructReportGated: Input active = false, finger_id = 0, button = 1, counter = 29001
VoodooInputSimulatorDevice::constructReportGated: Input active = true, finger_id = 1, button = 0, counter = 29002
VoodooInputSimulatorDevice::constructReportGated: enter, counter = 29003
`

Goshin commented:
Thanks for the info. It seems now we have a deeper insight into the problem.
Have you released the second finger at time=1544?
```
VoodooInputSimulatorDevice::constructReportGated: enter, time = 1544 
VoodooInputSimulatorDevice Input active is not detected at 0, button = 1 
VoodooInputSimulatorDevice Input active is detected at 1, button = 0
```
I cannot distinguish the second finger retouching because there are always two fingers in the log. Plus it would be better to also log the `x`, `y` position values, so we can tell whether the finger ids match the fingers.

Btw, if you are using VoodooI2C+VoodooInput by @kprinssu 's PR https://github.com/alexandred/VoodooI2C/pull/255, take a look at the bug in this comment.
https://github.com/alexandred/VoodooI2C/pull/255#issuecomment-601684782

lvs1974 commented:
@Goshin, I use VoodooI2C by @kprinssu and latest VoodooInput from acidanthera.
I disabled my patches in VoodooInput and added cleaning up of message in VoodooI2CNativeEngine::handleInterruptReport:
`message = {};
  message.timestamp = timestamp;
  message.contact_count = event.contact_count;`
But it did not help me, I see the same issues.
Later I will add more logging with finger id and x, y values.
If you need more info - please feel free to post here a code snippet which I can use to take logs.

lvs1974 commented:
@Goshin, it is hard to match my movements and events in the log. I attached this log, it stops when selection was unexpectedly aborted. I printed out coordinates and pressure, but transducer->currentCoordinates.pressure is always 0 for some reason.
[dmesg.txt](https://github.com/acidanthera/VoodooInput/files/4363358/dmesg.txt)
My logging snippet:
`		IOLog("%s: transducer index = %d, phys_button = %u, isActive = %u, finger_id = %u, pressure = %u, x = %u, y = %u, width = %u\n", getName(), i, transducer->isPhysicalButtonDown, transducer->isTransducerActive, finger_id, transducer->currentCoordinates.pressure, transducer->currentCoordinates.x, transducer->currentCoordinates.y, transducer->currentCoordinates.width);`

Goshin commented:
Don't worry, I can recognize the movements from the log now because the fix does reduce the redundant confusing finger. It is normal and okay to get a 0 pressure in `transducer->currentCoordinates.pressure` because VoodooI2C cannot obtain pressure value from the device.

Forget about the finger id. 😂
@williambj1 reminds me it could be related to the issue https://github.com/alexandred/VoodooI2C/issues/208.
In his case, the separate physical button clicking doesn't act as a finger touch, so the MT2Simulator sends a "no finger" report, which is similar to your problem.

Check out the simple fix here, it solves @williambj1 's problem. hope it helps yours too.
https://github.com/Goshin/VoodooInput/commit/e99278fe28b2cb0d4e434fcbd84c00ada2443c59

lvs1974 commented:
@Goshin, I built VoodooInput from your branch (https://github.com/Goshin/VoodooInput/commits/fix-separate-button-bug) with a fix "Improve dragging with separate button", but[ I got even more stranger behaviour](https://drive.google.com/file/d/1Y1bfc21vYed6AL8XFhrWw2rfz0MsfbYP/view?usp=sharing) than before.
The only thing I did on this video was pressing left button and moving finger from high to low border of touchpad.

Goshin commented:
Oops, XD. From the log you attached above, I can see the first finger drifts when the second finger re-contacts, which is pretty weird. You can validate it with the App "FingerMgmt".

Plz Try this one:
https://github.com/Goshin/VoodooInput/commit/84183e8a424649f999d33f32549804173bb3e2fe
I revert something so that the state of the first finger will not be kept, but if it turns out to come back to the original problem, thing is going to be tricky in this way as we might have to ignore the first finger completely.

lvs1974 commented:
@Goshin, thank you! The version built with commit "Revert to resetting state with separate button" works like a charm, I think quite similar to mine fix.

Goshin commented:
> @Goshin, thank you! The version built with commit "Revert to resetting state with separate button" works like a charm, I think quite similar to mine fix.


Yeah sure, this two-line fix can be considered as the simplified version of your fix.

https://github.com/acidanthera/VoodooInput/compare/master...Goshin:fix-separate-button-bug
```diff
@@ -79,7 +79,7 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
    input_report.timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;

    // finger data
-   bool input_active = false;
+   bool input_active = input_report.Button;
    bool is_error_input_active = false;

    for (int i = 0; i < multitouch_event.contact_count + 1; i++) {
@@ -203,7 +203,7 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
            finger_data.Pressure = 120;
        }

-       if (!transducer->isTransducerActive) {
+       if (!transducer->isTransducerActive && !input_report.Button) {
            finger_data.State = 0x7;
            finger_data.Priority = 0x5;
            finger_data.Size = 0x0;
```

You can go ahead with the PR.

lvs1974 commented:
@Goshin, thank you for the final fix. I still have a couple of off-topic questions, related to touchpad, and you probably could point me out in the right direction:

1. Quite often simple touchpad click (without any other fingers) does not work after 2-finger scrolling or simple 1-finger movement. I have to click 2-3 times to have click working. It happens in the left/right-bottom zone, but the center of touchpad (which is also clickable) works reliably.
2. In interrupt mode, in idle mode I have quite low cpu usage (~0.5% for kernel_task). If I keep one finger touched - I see 33-40% cpu usage (the same for simple movement), If I keep two finger pressed - I see 66% cpu usage. I tried to comment out everything in VoodooI2CHIDDevice::interruptOccured, but it did not help. I have no clue what does actually load CPU.
3. If I won't manage to get a good performance in interrupt mode (and I have many doubts), I would like to implement a kind of adaptive polling mode. It means we can use one time interval for idle mode (when touchpad is not use) and another interval for busy mode, for example 30ms and 3ms. I know it works brilliantly (and in idle I have cpu usage 1.5%, using touchpad - 7-8%). The only question is - how to detect whether touchpad is used. For my touchpad I can check return_size in VoodooI2CHIDDevice::getInputReport, if it is >= 14, that means touchpad is used. But may be you could suggest a more universal way? May be I could analyse some special flag set in VoodooInput, or something like that?

Thank you in advance!

alexandred commented:
> 2. In interrupt mode, in idle mode I have quite low cpu usage (~0.5% for kernel_task). If I keep one finger touched - I see 33-40% cpu usage (the same for simple movement), If I keep two finger pressed - I see 66% cpu usage. I tried to comment out everything in VoodooI2CHIDDevice::interruptOccured, but it did not help. I have no clue what does actually load CPU.

This is a bug that I believe I reduced down to either the I2C controller code (so somewhere in VooodooI2CControllerDriver iirc) or VoodooGPIO. We were unable to find the source of the issue inside there though. If you run the standard thread analysis tools in terminal you will see that the controller code is spinning up a ridiculously large amount of threads to deal with it’s interrupts (and thus a ridiculously large amount of threads on VoodooGPIO). I seem to recall now that devices that have ACPI controllers instead of PCI (so Haswell and Broadwell era machines) didn’t have this issue so it could indeed be related to using GPIO interrupts and would also explain why the issue doesn’t exist in polling mode (however I could be misremembering this)

Try having a looking at the threads - maybe you will spot something we missed. It’s indeed a very annoying and persistent bug and desperately needs fixing at some point.

lvs1974 commented:
> Try having a looking at the threads - maybe you will spot something we missed. It’s indeed a very annoying and persistent bug and desperately needs fixing at some point.

@alexandred: Thank you, I will try.

lvs1974 commented:
@alexandred, @Goshin, @kprinssu:  I think I have found something related to interrupts performance. The only thing I did - commented out the following lines, since these groups are not used in my touchpad:
```
static struct intel_community cnlh_communities[] = {
    CNLH_COMMUNITY(0, 0, 50, cnlh_community0_gpps), // Groups A and B
//    CNLH_COMMUNITY(1, 51, 154, cnlh_community1_gpps),
//    CNLH_COMMUNITY(2, 155, 248, cnlh_community3_gpps),
//    CNLH_COMMUNITY(3, 249, 298, cnlh_community4_gpps),
};
```

My calculations are below:
hexadecimal APIC pin number = 0x33
GPP_B3_IRQ = 0x33
GPP_B3	      = 28
GPP identifier = B
    CNL_GPP(1, 25, 50, 32),             /* GPP_B */
	      num, base, end, gpio_base
28 - 25 (base) + 32 (gpio_base) = 35 = 0x23

Now i have 18,5% for one touched finger, and 33,3% for two touched fingers.
When I excluded unused by me **GPP_A** from **cnlh_community0_gpps**, I have:
17,2% for one touched finger, and 30,8% for two fingers (this is instead of 33-40% and 66% in original version).
Could you have a look, please?

usr-sse2 commented:
> @williambj1 reminds me it could be related to the issue [alexandred/VoodooI2C#208](https://github.com/alexandred/VoodooI2C/issues/208).
> In his case, the separate physical button clicking doesn't act as a finger touch, so the MT2Simulator sends a "no finger" report, which is similar to your problem.

@Goshin 
In VoodooPS2SynapticsTouchpad there are two devices – VoodooInput's MT2 simulator and the old trackpad, and physical buttons (when not remapped into Force Touch) are sent using the old device, so finger touch simulation is not required.

Goshin commented:
> In VoodooPS2SynapticsTouchpad there are two devices – VoodooInput's MT2 simulator and the old trackpad, and physical buttons (when not remapped into Force Touch) are sent using the old device, so finger touch simulation is not required.

@usr-sse2 thanks for the info. I check out the related code, and it seems to be the `dispatchRelativePointerEvent` method and the `IOHIPointing` device. So we need to inherit `IOHIPointing` (as `ApplePS2SynapticsTouchPad` does) or to maintain an instance for it? That might be a little bit trickier in my view than sending a static spoofing packet for button click to the MT2Simulator.


> 1. Quite often simple touchpad click (without any other fingers) does not work after 2-finger scrolling or simple 1-finger movement. I have to click 2-3 times to have click working. It happens in the left/right-bottom zone, but the center of touchpad (which is also clickable) works reliably.

@lvs1974 The MT2Simulator spoofs the pressure value on the finger touch to simulate the button click, as there is no physical button on Magic Trackpad,

https://github.com/acidanthera/VoodooInput/blob/ec9d3dedc2a3fbc142bf87a5dbc67d7d5ec2249e/VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp#L200-L202

, and you can see currently in this process a valid touch event is needed. The problem is that the physical buttons of generic touchpad work independently to the finger touch, and the implementation and corporation of button and finger touch are device-specific. You can check out the details in [VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp](https://github.com/alexandred/VoodooI2CHID/blob/ee2620720850a652eb493dbd377da16c1fd57fef/VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp).

So you may want to check the touch log about the button pressing and releasing. From the log attached so far, I can see there is instability as the `transducer->isTransducerActive` occasionally shifts to 0 when the finger is still there, and the finger position drifts when another finger comes in.

If your touchpad reports the button state together with the touch position in `VoodooI2CMultitouchHIDEventDriver::handleDigitizerTransducerReport`, you can try the approach mentioned in @usr-sse2 's comment.

Goshin commented:
> @alexandred, @Goshin, @kprinssu: I think I have found something related to interrupts performance. The only thing I did - commented out the following lines, since these groups are not used in my touchpad:
> 
> ```
> static struct intel_community cnlh_communities[] = {
>     CNLH_COMMUNITY(0, 0, 50, cnlh_community0_gpps), // Groups A and B
> //    CNLH_COMMUNITY(1, 51, 154, cnlh_community1_gpps),
> //    CNLH_COMMUNITY(2, 155, 248, cnlh_community3_gpps),
> //    CNLH_COMMUNITY(3, 249, 298, cnlh_community4_gpps),
> };
> ```
> 
> My calculations are below:
> hexadecimal APIC pin number = 0x33
> GPP_B3_IRQ = 0x33
> GPP_B3 = 28
> GPP identifier = B
> CNL_GPP(1, 25, 50, 32), /* GPP_B */
> num, base, end, gpio_base
> 28 - 25 (base) + 32 (gpio_base) = 35 = 0x23
> 
> Now i have 18,5% for one touched finger, and 33,3% for two touched fingers.
> When I excluded unused by me **GPP_A** from **cnlh_community0_gpps**, I have:
> 17,2% for one touched finger, and 30,8% for two fingers (this is instead of 33-40% and 66% in original version).
> Could you have a look, please?

I think the problem might have something to do with the `VoodooGPIO::InterruptOccurred`, `VoodooGPIO::interruptOccurredGated`, and `VoodooGPIO::intel_gpio_community_irq_handler`.
